controllers/cache_controller.go

```diff
@@ -40,144 +61,399 @@ func (r *CacheReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		err = fmt.Errorf("unable to obtain default cache template: %w", err)
+		r.reqLogger.Error(err, "Error getting default XML")
+		return err
+	}
+
+	err = cluster.CreateCacheWithConfiguration(r.cache.Spec.Name, template, podName, mime.ApplicationXml)
 	if err != nil {
-		return reconcile.Result{}, err
+		err = fmt.Errorf("unable to create cache using default template: %w", err)
+		r.reqLogger.Error(err, "Error in creating cache")
+		return err
 	}
-	existsCache, err := cluster.ExistsCache(instance.GetCacheName(), podList.Items[0].Name)
-	if err == nil {
-		if existsCache {
-			reqLogger.Info(fmt.Sprintf("Cache %s already exists", instance.GetCacheName()))
-			// Check if template matches?
+	return nil
+}
+
+func (r *cacheRequest) reconcileDataGrid(cacheExists bool, podName string, cluster infinispan.ClusterInterface) error {
+	spec := r.cache.Spec
+	cacheName := r.cache.GetCacheName()
+	if cacheExists {
+		if spec.TemplateName != "" {
+			// TODO enforce with webhook validation when supported
+			r.log.Error(fmt.Errorf("updating an existing Cache's 'spec.TemplateName' field is not supported"), "")
 		} else {
-			reqLogger.Info(fmt.Sprintf("Cache %s doesn't exist, create it", instance.GetCacheName()))
-			podName := podList.Items[0].Name
-			templateName := instance.Spec.TemplateName
-			if ispnInstance.Spec.Service.Type == infinispanv1.ServiceTypeCache && (templateName != "" || instance.Spec.Template != "") {
-				errTemplate := fmt.Errorf("cannot create a cache with a template in a CacheService cluster")
-				reqLogger.Error(errTemplate, "Error creating cache")
-				return reconcile.Result{}, err
+			err := cluster.UpdateCacheWithConfiguration(cacheName, spec.Template, podName, mime.GuessMarkup(spec.Template))
+			if err != nil {
+				return fmt.Errorf("unable to update cache template: %w", err)
 			}
-			if templateName != "" {
-				err = cluster.CreateCacheWithTemplateName(instance.Spec.Name, templateName, podName)
-				if err != nil {
-					reqLogger.Error(err, "Error creating cache with template name")
-					return reconcile.Result{}, err
+		}
+		return nil
+	}
+
+	var err error
+	if spec.TemplateName != "" {
+		if err = cluster.CreateCacheWithTemplateName(cacheName, spec.TemplateName, podName); err != nil {
+			err = fmt.Errorf("unable to create cache with template name '%s': %w", spec.TemplateName, err)
+		}
+	} else {
+		if err = cluster.CreateCacheWithConfiguration(cacheName, spec.Template, podName, mime.GuessMarkup(spec.Template)); err != nil {
+			err = fmt.Errorf("unable to create cache with template: %w", err)
+		}
+	}
+
+	if err != nil {
+		r.reqLogger.Error(err, "Unable to create Cache")
+	}
+	return err
+}
+
+func (cl *CacheListener) CreateOrUpdate(data []byte) error {
+	cacheName, configYaml, err := unmarshallEventConfig(data)
+	if err != nil {
+		return err
+	}
+
+	if strings.HasPrefix(cacheName, "___") {
+		cl.Log.Debugf("Ignoring internal cache %s", cacheName)
+		return nil
+	}
+
+	cacheCrName := strcase.ToKebab(cacheName)
```

Comment:
As k8s doesn't support camelCase, I think we have two options here:

1. Existing code. Automatically convert camelCase to kebab-case. This will create a conflict  if the user has a definition for both "my-cache" and "myCache", "mycache" will be it's own CR. We can highlight this limitation in the documentation and hope the user reads it :slightly_smiling_face:

2. Don't support caches with camelCase names. In such a scenario, the only way to feedback to the user is via the config listener pod logs which isn't very user friendly.

WDYT? Any other ideas?

Reply:
Couldn't we use `spec.name` when a user really needs a camelCase name?

tbh, I think your current implementation+ some docs is a good enough solution. User that creates `my-cache` and `myCache` is probably looking for troubles :)
so it's ok to keep code as is for now.

Reply:
> Couldn't we use spec.name when a user really needs a camelCase name?

We already set `spec.Name` to the value received from the server to ensure that the HTTP calls work as expected. The problem is at the k8s resource level. We could always try to name the k8s resource something like `cache-name + <identifier>` but then you have a problem if the user creates a cache on the server with the name `cacheName + <identifier>` :scream:.

Reply:
Oh right, I see. Ok to highlight this in the doc then.